subsys/bluetooth/audio/hap_harc.c

```diff
@@ -0,0 +1,1533 @@
+/*
+ * Copyright (c) 2023 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/audio/csip.h>
+#include <zephyr/bluetooth/audio/hap.h>
+#include <zephyr/bluetooth/audio/has.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/check.h>
+
+#include "../bluetooth/host/hci_core.h"
+#include "csip_internal.h"
+#include "has_internal.h"
+
+#include "common/bt_str.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_hap_harc, CONFIG_BT_HAP_LOG_LEVEL);
+
+enum harc_flags {
+	FLAG_CONNECTED,
+	FLAG_PENDING_PROC,
+	FLAG_PENDING_CSIP_DISCOVERY,
+
+	FLAG_NUM,
+};
+
+enum harc_proc_flag {
+	PROC_FLAG_BUSY,
+	PROC_FLAG_PENDING_LOCK,
+	PROC_FLAG_PENDING_RELEASE,
+	PROC_FLAG_PENDING_OAP,
+	PROC_FLAG_LOCKED,
+	PROC_FLAG_NUM,
+};
+
+typedef int (*harc_proc_func_t)(struct bt_hap_harc *harc, void *user_data);
+
+struct harc_proc {
+	/** Procedure opcode */
+	uint8_t opcode;
+
+	/** Internal flags */
+	ATOMIC_DEFINE(flags, PROC_FLAG_NUM);
+
+	/** Array of controller instances to perform procedure on. */
+	struct bt_hap_harc *harcs[BT_HAP_HARC_SET_SIZE_MAX];
+
+	/** Instance count */
+	uint8_t count;
+
+	/** Set information deep-copy (in case of CSIP pointer is removed) */
+	struct bt_csip_set_coordinator_set_info set_info;
+
+	/** Pending procedure user data */
+	void *user_data;
+};
+
+static struct csip_set_member {
+	/** Set member object pointer */
+	const struct bt_csip_set_coordinator_set_member *member;
+
+	/** CSIS instance pointer. May be NULL */
+	const struct bt_csip_set_coordinator_csis_inst *csis;
+
+	/** The number of sets the object is member of */
+	size_t set_count;
+} set_members[CONFIG_BT_MAX_CONN];
+
+static struct bt_hap_harc {
+	/** HAS client instance */
+	struct bt_has_client *client;
+
+	/** Set identifier */
+	uint8_t set_id;
```

Comment:
I'm a bit confused about this. 

What is a set ID? 
Why is the conn_index being assigned it? Isn't that pretty much the opposite of a set ID?

Reply:
It's used to pair the devices. So the devices with the same `set_id` value are known to be in set.
The devices that are not in set will just have `set_id` value == bt_conn_index.

Reply:
Maybe it's just because I can't find it in the GH code browser, but I don't see `set_id` being assigned to anything but `harc->set_id = conn_index;` in `bt_hap_harc_bind` and then with `harc_1->set_id = harc_0->set_id;` in `harc_set_add`, but if `harc_0->set_id` is always the `conn_index` then it's equally confusing.